### PR TITLE
*: DDL drop database/table

### PR DESCRIPTION
**Pull request by zimulala**
#800 
Handle to the first three `TODO`

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @qiuyesuifeng @siddontang

File: ddl/ddl.go [hunk 1]
```diff
@@ -75,6 +75,8 @@ type ddl struct {
 	uuid      string
 	jobCh     chan struct{}
 	jobDoneCh chan struct{}
+	// drop database/table task
+	taskCh chan struct{}
```
Comment (shenli) [inline ddl/ddl.go]: If the chan is only used for drop task. So it is better to rename it to dropTaskCh.

File: ddl/ddl.go [hunk 2]
```diff
@@ -121,16 +124,26 @@ func (d *ddl) Stop() error {
...
+		if err != nil {
 			return errors.Trace(err)
 		}
-
-		if owner.OwnerID != d.uuid {
+		if owner == nil || owner.OwnerID != d.uuid {
 			return nil
 		}
-
 		// owner is mine, clean it so other servers can compete for it quickly.
```
Comment (shenli) [inline ddl/ddl.go]: Owner is me

File: ddl/ddl.go (hunk 1 above)
Comment (zimulala) [inline ddl/ddl.go]: yep, it‘s better.

File: ddl/ddl.go (hunk 2 above)
Comment (zimulala) [inline ddl/ddl.go]: Gotcha.

File: ddl/ddl.go [hunk 3]
```diff
@@ -121,16 +124,26 @@ func (d *ddl) Stop() error {
...
-		return errors.Trace(t.SetDDLOwner(&model.Owner{}))
+		if err = t.SetDDLOwner(&model.Owner{}); err != nil {
+			return errors.Trace(err)
+		}
+
+		// clean task owner like job owner.
```
Comment (shenli) [inline ddl/ddl.go]: The comment is not clear.

File: ddl/ddl.go [hunk 4]
```diff
@@ -121,16 +124,26 @@ func (d *ddl) Stop() error {
...
+		owner, err = t.GetDDLTaskOwner()
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: Should DDL and DDL Task owner be the same one?

Comment (zimulala) [inline ddl/ddl.go]: no, it shouldn't. The two owners aren't the same to make it real executed asynchronously.

File: ddl/worker.go [hunk 5]
```diff
@@ -158,6 +178,13 @@ func (d *ddl) finishJob(t *meta.Meta, job *model.Job) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+	switch job.Type {
+	case model.ActionDropSchema, model.ActionDropTable:
+		//	model.ActionDropColumn, model.ActionDropIndex:
```
Comment (qiuyesuifeng) [inline ddl/worker.go]: Remove this line.

Comment (zimulala) [inline ddl/worker.go]: got it.

File: meta/meta.go [hunk 6]
```diff
@@ -476,15 +491,26 @@ func (m *Meta) GetDDLJob(index int64) (*model.Job, error) {
 	return job, errors.Trace(err)
 }
 
-// UpdateDDLJob updates the DDL job with index.
-func (m *Meta) UpdateDDLJob(index int64, job *model.Job) error {
+// GetDDLJob returns the DDL job with index.
+func (m *Meta) GetDDLJob(index int64) (*model.Job, error) {
+	job, err := m.getDDLJob(mDDLJobListKey, index)
+
```
Comment (qiuyesuifeng) [inline meta/meta.go]: Remove this empty line?

File: ddl/executor.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	err := t.AddHistoryDDLTask(task)
+
+	return errors.Trace(err)
+}
+
+func (d *ddl) onExecute() {
+	defer d.wait.Done()
+
+	// ensure that have ddl job convert to ddl task
+	checkTime := chooseLeaseTime(8*d.lease, 10*time.Second)
```
Comment (qiuyesuifeng) [inline ddl/executor.go]: Why choose lease time as `8*d.lease` here?

File: ddl/executor.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+	ticker := time.NewTicker(checkTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			log.Debugf("[ddl] wait %s to check DDL task status again", checkTime)
+		case <-d.taskCh:
+		case <-d.quitCh:
```
Comment (qiuyesuifeng) [inline ddl/executor.go]: onExecute and onWorker both wait from quitCh, then here may not work well.

File: ddl/executor_test.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	defer store.Close()
+
+	lease := 10 * time.Millisecond
+	d := newDDL(store, nil, nil, lease)
+	defer d.close()
+
+	task := &model.Job{
+		SchemaID: 1,
+		Type:     model.ActionDropSchema,
+		Args: []interface{}{&model.DBInfo{
```
Comment (qiuyesuifeng) [inline ddl/executor_test.go]: Can this pass gofmt?

File: ddl/executor.go (hunk 7 above)
Comment (zimulala) [inline ddl/executor.go]: Now the DDL flow for drop database/table is public -> write only -> delete only(include delete all data) -> absent.
I want to make sure the state is absent. But I'm not sure if this is the most appropriate time。

File: ddl/executor.go (hunk 8 above)
Comment (zimulala) [inline ddl/executor.go]: here we use `close(quitCh)`, so the two will work.

File: ddl/executor_test.go (hunk 9 above)
Comment (zimulala) [inline ddl/executor_test.go]: yep, it can pass make.

**Comment by zimulala** [comment]
PTAL @qiuyesuifeng @shenli  @siddontang

**Comment by shenli** [comment]
LGTM

File: ddl/ddl.go [hunk 10]
```diff
@@ -150,11 +162,13 @@ func (d *ddl) Start() error {
 
 func (d *ddl) start() {
 	d.quitCh = make(chan struct{})
-	d.wait.Add(1)
+	d.wait.Add(2)
+	go d.onExecute()
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: How about using onTaskWorker and onJobWorker?

File: ddl/ddl.go [hunk 11]
```diff
@@ -150,11 +162,13 @@ func (d *ddl) Start() error {
...
 	go d.onWorker()
 	// for every start, we will send a fake job to let worker
 	// check owner first and try to find whether a job exists and run.
 	asyncNotify(d.jobCh)
+	asyncNotify(d.dropTaskCh)
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: dropTaskCh seems confused, how about using backgroudTaskCh?

**Comment by siddontang** [comment]
JobFlag and TaskFlag seem confused, I thought they were the same thing.

**Comment by zimulala** [comment]
PTAL @siddontang @qiuyesuifeng

File: ddl/bg_worker.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// runBgJob runs background job.
+func (d *ddl) runBgJob(t *meta.Meta, task *model.Job) {
```
Comment (qiuyesuifeng) [inline ddl/bg_worker.go]: Please using `job *model.Job` as argument.

File: ddl/bg_worker.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// getFirstBgJob gets the first background job.
+func (d *ddl) getFirstBgJob(t *meta.Meta) (*model.Job, error) {
+	task, err := t.GetBgJob(0)
+	return task, errors.Trace(err)
+}
+
+// updateBgJob updates background job.
+func (d *ddl) updateBgJob(t *meta.Meta, task *model.Job) error {
```
Comment (qiuyesuifeng) [inline ddl/bg_worker.go]: Same here.

File: ddl/bg_worker.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+	err := t.AddHistoryBgJob(task)
+
+	return errors.Trace(err)
+}
+
+func (d *ddl) onBackgroundWorker() {
+	defer d.wait.Done()
+
+	// ensure that have ddl job convert to background job.
```
Comment (qiuyesuifeng) [inline ddl/bg_worker.go]: This comment is not clear.

File: ddl/ddl.go [hunk 15]
```diff
@@ -120,18 +123,35 @@ func (d *ddl) Stop() error {
...
+		if err1 != nil {
+			return errors.Trace(err1)
+		}
+		if owner == nil || owner.OwnerID != d.uuid {
 			return nil
 		}
 
-		// owner is mine, clean it so other servers can compete for it quickly.
-		return errors.Trace(t.SetDDLOwner(&model.Owner{}))
+		// task's owner is me, clean it so other servers can compete for it quickly.
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: background job's owner

File: ddl/ddl_worker.go [hunk 16]
```diff
@@ -105,8 +105,18 @@ func asyncNotify(ch chan struct{}) {
 	}
 }
 
-func (d *ddl) checkOwner(t *meta.Meta) (*model.Owner, error) {
-	owner, err := t.GetDDLOwner()
+func (d *ddl) checkOwner(t *meta.Meta, flag string) (*model.Owner, error) {
```
Comment (qiuyesuifeng) [inline ddl/ddl_worker.go]: Why not using an enum value for flag?

File: ddl/ddl_worker.go [hunk 17]
```diff
@@ -122,42 +132,58 @@ func (d *ddl) checkOwner(t *meta.Meta) (*model.Owner, error) {
 	// the owner will update its owner status every 2 * lease time, so here we use
 	// 4 * lease to check its timeout.
 	maxTimeout := int64(4 * d.lease)
+	if flag == bgJobFlag {
+		// backgroun job doesn't need to guarantee other servers update the schema.
```
Comment (qiuyesuifeng) [inline ddl/ddl_worker.go]: background

File: ddl/schema.go [hunk 18]
```diff
@@ -95,52 +118,28 @@ func (d *ddl) onDropSchema(t *meta.Meta, job *model.Job) error {
...
-			return nil
-		}
-		if err != nil {
-			return errors.Trace(err)
-		}
-
 		// all reorganization jobs done, drop this database
 		if err = t.DropDatabase(dbInfo.ID); err != nil {
-			return errors.Trace(err)
+			break
```
Comment (qiuyesuifeng) [inline ddl/schema.go]: Why here using break?

File: ddl/schema_test.go [hunk 19]
```diff
@@ -59,26 +59,49 @@ func testDropSchema(c *C, ctx context.Context, d *ddl, dbInfo *model.DBInfo) *mo
...
+	time.Sleep(5 * time.Millisecond)
+	return false
+}
+
 func testCheckSchemaState(c *C, d *ddl, dbInfo *model.DBInfo, state model.SchemaState) {
-	kv.RunInNewTxn(d.store, false, func(txn kv.Transaction) error {
-		t := meta.NewMeta(txn)
-		info, err := t.GetDatabase(dbInfo.ID)
-		c.Assert(err, IsNil)
+	isDrop := true
```
Comment (qiuyesuifeng) [inline ddl/schema_test.go]: Using false as default value?

Comment (qiuyesuifeng) [inline ddl/schema_test.go]: And how about dropped?

File: ddl/table.go [hunk 20]
```diff
@@ -76,6 +76,35 @@ func (d *ddl) onCreateTable(t *meta.Meta, job *model.Job) error {
...
 }
 
+func (d *ddl) delReorgTable(t *meta.Meta, task *model.Job) error {
+	// reorganization -> absent
+	tblInfo := &model.TableInfo{}
+	err := task.DecodeArgs(tblInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if tblInfo == nil {
```
Comment (qiuyesuifeng) [inline ddl/table.go]: Seems tblInfo will always be not nil?

File: ddl/ddl_worker.go (hunk 17 above)
Comment (qiuyesuifeng) [inline ddl/ddl_worker.go]: @siddontang What is your opinion of `maxTimeout` value here? Seems `2 * d.lease` is too long.

File: ddl/ddl_worker.go (hunk 16 above)
Comment (zimulala) [inline ddl/ddl_worker.go]: it‘s also used to log.

File: ddl/schema.go (hunk 18 above)
Comment (zimulala) [inline ddl/schema.go]: it’s consistent with the other parts.

File: ddl/bg_worker.go [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+	job.State = model.JobRunning
+
+	var err error
+	switch job.Type {
+	case model.ActionDropSchema:
+		err = d.delReorgSchema(t, job)
+	case model.ActionDropTable:
+		err = d.delReorgTable(t, job)
+	default:
+
```
Comment (qiuyesuifeng) [inline ddl/bg_worker.go]: Remove empty line.

File: ddl/ddl.go [hunk 22]
```diff
@@ -71,10 +71,12 @@ type ddl struct {
...
 	// schema lease seconds.
-	lease     time.Duration
-	uuid      string
-	jobCh     chan struct{}
-	jobDoneCh chan struct{}
+	lease        time.Duration
+	uuid         string
+	ddlJobCh     chan struct{}
+	ddlJobDoneCh chan struct{}
+	// drop database/table job runs in the background to preform.
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: to preform?

File: ddl/ddl_worker.go (hunk 16 above)
Comment (qiuyesuifeng) [inline ddl/ddl_worker.go]: Then u can implement a String interface like we do in JoinType.

File: ddl/ddl_worker.go [hunk 23]
```diff
@@ -122,42 +132,58 @@ func (d *ddl) checkOwner(t *meta.Meta) (*model.Owner, error) {
 	// the owner will update its owner status every 2 * lease time, so here we use
 	// 4 * lease to check its timeout.
 	maxTimeout := int64(4 * d.lease)
+	if flag == bgJobFlag {
+		// background job doesn't need to guarantee other servers update the schema.
+		maxTimeout = int64(2 * d.lease)
```
Comment (qiuyesuifeng) [inline ddl/ddl_worker.go]: If u need not guarantee the other servers update the schema, why not using a short time?
`2 * d.lease` is a little long.

File: ddl/table.go [hunk 24]
```diff
@@ -76,6 +76,33 @@ func (d *ddl) onCreateTable(t *meta.Meta, job *model.Job) error {
...
+	// reorganization -> absent
+	tblInfo := &model.TableInfo{}
+	err := job.DecodeArgs(tblInfo)
+	if err != nil {
+		// arg error, cancel this job.
+		job.State = model.JobCancelled
+		return errors.Trace(err)
+	}
+	tblInfo.State = model.StateDeleteReorganization
+	tbl, err := d.getTable(job.SchemaID, tblInfo)
```
Comment (qiuyesuifeng) [inline ddl/table.go]: Seems u have called DropTable in onDropTable function, then could u get table info here?

File: ddl/ddl.go (hunk 22 above)
Comment (zimulala) [inline ddl/ddl.go]: I will remove `to preform`.

File: ddl/table.go (hunk 24 above)
Comment (zimulala) [inline ddl/table.go]: yep，I have put table info to job.

**Comment by qiuyesuifeng** [comment]
@siddontang PTAL

File: ddl/bg_worker.go [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+		if job.State != model.JobCancelled {
+			log.Errorf("run background job err %v", errors.ErrorStack(err))
+		}
+
+		job.Error = err.Error()
+		job.ErrorCount++
+	}
+}
+
+// prepareBgJob prepares background job.
```
Comment (ngaut) [inline ddl/bg_worker.go]: a background job

File: ddl/bg_worker.go [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+	})
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// runBgJob runs background job.
```
Comment (ngaut) [inline ddl/bg_worker.go]: a background job

File: ddl/bg_worker.go [hunk 27]
```diff
@@ -0,0 +1,178 @@
...
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/terror"
+)
+
+// handleBgJobQueue handles background job queue.
```
Comment (ngaut) [inline ddl/bg_worker.go]: a background job

File: ddl/bg_worker_test.go [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+		}},
+	}
+	d.prepareBgJob(job)
+	d.startBgJob(job.Type)
+
+	time.Sleep(lease)
+	testCheckBgJobState(c, d, job, model.JobDone)
+}
+
+func testCheckBgJobState(c *C, d *ddl, job *model.Job, state model.JobState) {
```
Comment (ngaut) [inline ddl/bg_worker_test.go]: s/testCheckBgJobState/verifyBgJobState

File: meta/meta.go [hunk 29]
```diff
@@ -552,3 +586,65 @@ func (m *Meta) GetDDLReorgHandle(job *model.Job) (int64, error) {
...
+	return job, errors.Trace(err)
+}
+
+// EnQueueBgJob adds a background job to the list.
+func (m *Meta) EnQueueBgJob(job *model.Job) error {
+	return m.enQueueDDLJob(mBgJobListKey, job)
+}
+
+// BgJobLength returns the background job length.
+func (m *Meta) BgJobLength() (int64, error) {
```
Comment (ngaut) [inline meta/meta.go]: s/BgJobLength/BgJobQueueLength

**Comment by ngaut** [comment]
LGTM

**Pull request merged**